staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/validation.go

```diff
@@ -1014,6 +1015,9 @@ func ValidateCustomResourceDefinitionOpenAPISchema(schema *apiextensions.JSONSch
 			} else if hasNewlines(trimmedRule) && len(trimmedMsg) == 0 {
 				allErrs.SchemaErrors = append(allErrs.SchemaErrors, field.Required(fldPath.Child("x-kubernetes-validations").Index(i).Child("message"), "message must be specified if rule contains line breaks"))
 			}
+			if len(rule.MessageExpression) > 0 && len(trimmedMsgExpr) == 0 {
+				allErrs.SchemaErrors = append(allErrs.SchemaErrors, field.Required(fldPath.Child("x-kubernetes-validations").Index(i).Child("messageExpression"), "messageExpression must be non-empty if specified"))
+			}
```

Comment:
Couple things missed earlier: 
- Should we loose the previous validation on `message`?
```
else if len(rule.Message) > 0 && len(trimmedMsg) == 0 {
				allErrs.SchemaErrors = append(allErrs.SchemaErrors, field.Invalid(fldPath.Child("x-kubernetes-validations").Index(i).Child("message"), rule.Message, "message must be non-empty if specified"))
			} else if hasNewlines(trimmedMsg) {
				allErrs.SchemaErrors = append(allErrs.SchemaErrors, field.Invalid(fldPath.Child("x-kubernetes-validations").Index(i).Child("message"), rule.Message, "message must not contain line breaks"))
			} else if hasNewlines(trimmedRule) && len(trimmedMsg) == 0 {
				allErrs.SchemaErrors = append(allErrs.SchemaErrors, field.Required(fldPath.Child("x-kubernetes-validations").Index(i).Child("message"), "message must be specified if rule contains line breaks"))
```
  
- Should we add check on newlines in MessageExpression?

The first question might be no due to backward compatible concern.

Reply:
Makes sense, I'll add a line break check (I'm not sure `message must be specified if rule contains line breaks` is applicable to `messageExpression`).

Reply:
After offline discussion with @jpbetz , let's leave it as it is now since there is no control here on if cel program gonna return multiple lines string or one line. But to keep the behavior sync and avoid returning large message, we could consider:
- through runtime error when messageExpression returned multiple lines string
- add a length limit on the output(If the resulting message is more than 5k, it is not returned and instead the message is used)

Reply:
Sounds good, I'll add those two instead.

Reply:
(marking as resolved due to beb0f1c4ea5e78718584fe10295bb03de15147cb)